### PR TITLE
Add vault crypto utilities

### PR DESCRIPTION
**Pull request by curleyr**
**Description:**
- This PR Introduces client-side cryptography utilities and corresponding unit tests that enable secure encryption and decryption of vault entries before transmission to or from the backend. These utilities ensure the server never has access to plaintext vault data, supporting a zero-knowledge encryption design.

**Features:**
1. AES-GCM Encryption Utilities
    - Implements encryptVaultEntry and decryptVaultEntry using the Web Crypto API.
    - Uses AES-GCM (256-bit) for authenticated encryption with randomly generated 96-bit nonces.
    - Ciphertext and nonce are packed together into a single hex-encoded string for compact storage:
      - Encryption: Nonce is prepended to ciphertext.
      - Decryption: Nonce is extracted from first 12 bytes of ciphertext.

    Example usage:
    ```javascript
    const vaultKey = '06cf04db65c36c16cacc92a657de6bc0729da9e6b73c4342bf842bf58c434b48'; // 256-bit key derived on server
    const plaintext = 'P@ssw0rd!'; // secret value never visible to the server

    // Encrypt before sending to server (for storage)
    const ciphertext = await encryptVaultEntry(vaultKey, plaintext);

    // Decrypt after retrieving from server (for client-side display)
    const decrypted = await decryptVaultEntry(vaultKey, ciphertext);
    ```

2. Encoding Helpers
    - toBytes() converts strings, objects, or existing byte arrays to Uint8Array form.
    - toHex() converts bytes to hexadecimal string.
    - Provides validation and error handling.
    - See top of file comments for adapted code examples from:
      - [Stack Overflow](https://stackoverflow.com/a/69980864)
      - [Chat GPT](https://chatgpt.com/share/690672d1-b4cc-8001-aad7-28ccbb57e4d1)

**Testing and Coverage:**
- Added full test suite under VaultCryptoUtils.test.js using Vitest.
- Run commands:
  ```bash
  pnpm test        # run all tests
  pnpm coverage    # run tests with coverage report
  ```
- All tests passed with 100% coverage.
  ```bash
   ✓ src/tests/vault_crypto.test.js (16 tests) 8ms
     ✓ toBytes (5)
       ✓ leaves bytes as bytes 1ms
       ✓ encodes hex to bytes 1ms
       ✓ encodes regular (non-hex) string to bytes 0ms
       ✓ encodes JSON to bytes 0ms
       ✓ throws error for unsupported type 0ms
     ✓ toHex (4)
       ✓ leaves hex as hex 0ms
       ✓ converts bytes to hex 0ms
       ✓ throws error for non-hex string 0ms
       ✓ throws error for unsupported type 0ms
     ✓ encryptVaultEntry/decryptVaultEntry (7)
       ✓ encrypt/decrypt returns original plaintext (string) 3ms
       ✓ encrypt/decrypt returns original plaintext (object) 0ms
       ✓ random nonce produces different ciphertext each time 0ms
       ✓ decrypt fails with wrong key 1ms
       ✓ throws for unsupported vaultKey type 0ms
       ✓ throws for unsupported data type 0ms
       ✓ decrypt fails for malformed ciphertext (no nonce) 0ms
  
   Test Files  1 passed (1)
        Tests  16 passed (16)
     Start at  17:16:13
     Duration  169ms (transform 19ms, setup 0ms, collect 26ms, tests 8ms, environment 0ms, prepare 13ms)
  
   % Coverage report from v8
  -----------------|---------|----------|---------|---------|-------------------
  File             | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
  -----------------|---------|----------|---------|---------|-------------------
  All files        |     100 |      100 |     100 |     100 |                   
   vault_crypto.js |     100 |      100 |     100 |     100 |                   
  -----------------|---------|----------|---------|---------|-------------------
  ```

### CODE REVIEW

**Review by beckywong37 — COMMENTED** [review]
Hi Bobby,
Great work on the PR for the crypto utility module! 
- Good use of the Web Crypto API for AES-GCM and Nonce
- Good documentation and references, including comments in the code and at the top of the file
- Tests are thorough with 100% coverage. Edge cases such as unsupported types and wrong key
- Proper type checking and input validation in helper utilities

One structural recommendation, especially since all the crypto functions are stateless, is to split the file into logical units. For example, in src/utils/vault_crypto, you may have 
```
- toBytes.ts
- toHex.ts
- encryptValutEntry.ts
- decryptVaultEntry.ts
```
This can help with visibility in large codebases since each method is a top-level file and it keeps the code more modular. 

In your `toHex() `and `toBytes()` function, it is calling the following:
`const hexRegex = /^[0-9a-fA-F]+$/;
`
Alternatively, the Regex can be declared at the top of the file as a constant so HEX_REGEX is created once when the module is loaded:

```
const HEX_REGEX = /^[0-9a-fA-F]+$/;

function toHex(value) {
    if (HEX_REGEX.test(value)) {
        // ...
    }
}

```

Your current `hex` implementation uses manual concatenation to build the string. Using the `Array.from` and `join(' ')` function can help make the code more readable and avoid repeated string concatenation in a loop.

```
if (value instanceof Uint8Array) {
  return Array.from(value, byte => byte.toString(16).padStart(2, '0')).join('');
}
```

Overall great work as always on the secure crypto implementation with comprehensive testing!

File: frontend/web/src/utils/vault_crypto.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+*/
+
+/**
+* Helper for UTF-8 encoding of text or JSON
+* See reference listed above
+* @param {string|Uint8Array|object} value
+* @returns {Uint8Array} 
+*/
+export function toBytes(value) {
+    if (value instanceof Uint8Array) return value;
```
Comment (xiaolongbytes) [inline frontend/web/src/utils/vault_crypto.js]: Great short-circuiting!

File: frontend/web/src/utils/vault_crypto.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+    if (typeof value === 'string') {
+        // Check if string is valid hex (even length and only hex chars)
```
Comment (xiaolongbytes) [inline frontend/web/src/utils/vault_crypto.js]: Could create a helper function like `is_valid_hex_string` to make this a bit more readable and abstract away the implementation details.

File: frontend/web/src/utils/vault_crypto.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+*     the latest versions of browsers. This can be revisited at a later date as they mature.
+*     
+*     https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array/toHex
+*     https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array/setFromHex
+*/
+
+/**
+* Helper for UTF-8 encoding of text or JSON
+* See reference listed above
+* @param {string|Uint8Array|object} value
```
Comment (xiaolongbytes) [inline frontend/web/src/utils/vault_crypto.js]: I love how you document the types in the param and returns. It instantly answers the question "what types does this convert to Bytes?"

File: frontend/web/src/utils/vault_crypto.js (hunk 2 above)
Comment (curleyr) [reply frontend/web/src/utils/vault_crypto.js]: Good idea. I'll create that when I separate the current .js file into separate .ts modules

File: frontend/web/src/utils/vault_crypto.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+* @returns {string}
+*/
+export function toHex(value) {
+    // Check if already valid hex
+    if (typeof value === 'string') {
+        const hexRegex = /^[0-9a-fA-F]+$/;
+        if (hexRegex.test(value) && value.length % 2 === 0) {
+            return value;
+        }
+        throw new TypeError(`Invalid hex string`);
```
[the remark this answers is not in the document]
Comment (curleyr) [reply frontend/web/src/utils/vault_crypto.js]: Done
